File: app/src/main/java/com/bakkenbaeck/sol/view/custom/SunView.java
package com.bakkenbaeck.sol.view.custom;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.view.View;
import com.bakkenbaeck.sol.R;

public class SunView extends View {
    private Paint paint;

    private String startLabel = "";
    private String endLabel = "";
    private String floatingLabel = "";

    private int bottom;
    private int left;
    private int right;
    private int width;
    private int textMargin;
    private int horizon_bottom_margin;
    private int circleRadius;
    private int availableSunHeight;
    private int availableSunWidth;
    private int strokeWidth;
    private int rightMarginLabel;
    private int leftMarginLabel;

    private double[] coords;
    private boolean sunIsVisible;

    public SunView(final Context context) {
        super(context);
        init();
    }

    public SunView(final Context context, final AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public SunView(final Context context, final AttributeSet attrs, final int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        this.paint = new Paint();
        this.paint.setStyle(Paint.Style.FILL);
        this.coords = new double[]{0.0, 0.0};

        initSizes();
    }

    public void initSizes() {
        this.paint.setColor(ContextCompat.getColor(getContext(), R.color.sun_default_color));
        this.paint.setTextSize(this.getContext().getResources().getDimensionPixelSize(R.dimen.sun_text_size));
        this.textMargin = this.getContext().getResources().getDimensionPixelSize(R.dimen.sun_text_margin);
        this.horizon_bottom_margin = this.getContext().getResources().getDimensionPixelSize(R.dimen.sun_horizon_bottom_margin);
        this.circleRadius = this.getContext().getResources().getDimensionPixelSize(R.dimen.sun_radius);
        this.strokeWidth = this.getContext().getResources().getDimensionPixelSize(R.dimen.sun_stroke_width);
        this.rightMarginLabel = getContext().getResources().getDimensionPixelSize(R.dimen.sun_right_margin_label);
        this.leftMarginLabel = getContext().getResources().getDimensionPixelSize(R.dimen.sun_left_margin_label);
        this.paint.setStrokeWidth(strokeWidth);
    }

    public void setStrokeWidth(final int strokeWidth) {
        this.paint.setStrokeWidth(strokeWidth);
    }

    public SunView setTypeface(final Typeface typeface) {
        this.paint.setTypeface(typeface);
        return this;
    }

    public SunView setColor(final int color) {
        this.paint.setColor(color);
        return this;
    }

    public SunView setStartLabel(final String startLabel) {
        this.startLabel = startLabel;
        return this;
    }

    public SunView setEndLabel(final String endLabel) {
        this.endLabel = endLabel;
        return this;
    }

    public SunView setFloatingLabel(final String floatingLabel) {
        this.floatingLabel = floatingLabel;
        return this;
    }

    public SunView setTextSize(final int textSize) {
        this.paint.setTextSize(textSize);
        return this;
    }

    public SunView setFloatingTextBottomMargin(final int textMargin) {
        this.textMargin = textMargin;
        return this;
    }

    public SunView setSunRadius(final int radius) {
        this.circleRadius = radius;
        return this;
    }

    public SunView setHorizonMargin(final int margin) {
        this.horizon_bottom_margin = margin;
        return this;
    }

    @Override
    protected void onSizeChanged(final int w, final int h, final int oldw, final int oldh) {
        this.width = w;
        this.bottom = h;
        this.left = 0;
        this.right = w;

        this.availableSunHeight = h - horizon_bottom_margin - textMargin;
        this.availableSunWidth = w - (circleRadius * 4);

        super.onSizeChanged(w, h, oldw, oldh);
    }

    @Override
    protected void onDraw(final Canvas canvas) {
        super.onDraw(canvas);

        drawLabels(canvas);
        drawHorizontal(canvas);
        clipCanvas(canvas);
        drawSun(canvas);
    }

    private void drawLabels(final Canvas canvas) {
        if (this.startLabel != null) {
            canvas.drawText(
                    this.startLabel,
                    this.left + leftMarginLabel,
                    this.bottom - 1,
                    this.paint);
        }

        if (this.endLabel != null) {
            final Rect bounds = new Rect();
            this.paint.getTextBounds(
                    this.endLabel,
                    0,
                    this.endLabel.length(),
                    bounds);
            canvas.drawText(
                    this.endLabel,
                    this.right - bounds.width() - rightMarginLabel,
                    this.bottom - 1,
                    this.paint);
        }
    }

    private void drawHorizontal(final Canvas canvas) {
        canvas.drawLine(
                this.left,
                this.bottom - horizon_bottom_margin,
                this.right,
                this.bottom - horizon_bottom_margin,
                this.paint);
    }

    private void drawSun(Canvas canvas) {
        drawCircle(canvas, coords);
        if (sunIsVisible) {
            drawFloatingLabel(canvas, coords);
        }
    }

    private void clipCanvas(final Canvas canvas) {
        canvas.clipRect(
                0,
                0,
                this.width,
                this.bottom - horizon_bottom_margin);
    }

    public SunView setPercentProgress(final double progress) {
        final double position = (Math.PI + (progress * Math.PI));
        double x = (50 + Math.cos(position) * 50) / 100;
        double y = (Math.abs(Math.sin(position) * 100)) / 100;

        if (progress > 1 || progress < 0) {
            x = 0;
            y = 0;
            sunIsVisible = false;
        } else {
            sunIsVisible = true;
        }

        this.coords = new double[]{x, y};
        invalidate();

        return this;
    }

    private void drawCircle(final Canvas canvas, final double[] coords) {
        canvas.drawCircle(
                (int)(coords[0] * availableSunWidth) + (circleRadius * 2),
                (int)(this.availableSunHeight - (this.availableSunHeight * coords[1]))
                        + circleRadius + textMargin,
                circleRadius,
                this.paint);
    }

    private void drawFloatingLabel(final Canvas canvas, final double[] coords) {
        if (this.floatingLabel != null) {
            final Rect bounds = new Rect();
            this.paint.getTextBounds(
                    this.floatingLabel,
                    0,
                    this.floatingLabel.length(),
                    bounds);
            canvas.drawText(
                    this.floatingLabel,
                    (int)((coords[0] * availableSunWidth) - (bounds.width() / 2)) + (circleRadius * 2),
                    (int)((this.availableSunHeight - (this.availableSunHeight * coords[1]))
                            - (bounds.height() * 2)) + circleRadius + textMargin,
                    this.paint);
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        final double height = getMeasuredWidth() * 0.6;
        final int newHeight = (int) height;
        setMeasuredDimension(getMeasuredWidth(), newHeight);
    }
}

File: app/src/main/java/com/bakkenbaeck/sol/view/custom/SunView.kt
package com.bakkenbaeck.sol.view.custom

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import androidx.core.content.ContextCompat
import com.bakkenbaeck.sol.R
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin

class SunView : View {

    private val paint by lazy {
        Paint().apply {
            style = Paint.Style.FILL
            color = ContextCompat.getColor(context, R.color.sun_default_color)
            textSize = resources.getDimensionPixelSize(R.dimen.sun_text_size).toFloat()
        }
    }

    private var startLabel: String = ""
    private var endLabel: String = ""
    private var floatingLabel: String = ""

    private val viewRect by lazy { Rect() }

    private var textMargin = 0f
    private var horizon_bottom_margin = 0
    private var circleRadius = 0
    private var availableViewWidth = 0
    private var strokeWidth = 0
    private var rightMarginLabel = 0f
    private var leftMarginLabel = 0f
    private val sunCoordinates: PointF by lazy { PointF(.0f, .0f) }
    private var sunIsVisible = false

    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    )

    init {
        val resources = context.resources
        textMargin = resources.getDimensionPixelSize(R.dimen.sun_text_margin).toFloat()
        horizon_bottom_margin = resources.getDimensionPixelSize(R.dimen.sun_horizon_bottom_margin)
        circleRadius = resources.getDimensionPixelSize(R.dimen.sun_radius)
        strokeWidth = resources.getDimensionPixelSize(R.dimen.sun_stroke_width)
        rightMarginLabel = resources.getDimensionPixelSize(R.dimen.sun_right_margin_label).toFloat()
        leftMarginLabel = resources.getDimensionPixelSize(R.dimen.sun_left_margin_label).toFloat()
        paint.strokeWidth = strokeWidth.toFloat()
    }

    fun setColor(color: Int) {
        paint.color = color
    }

    fun setStartLabel(value: String?) {
        startLabel = value ?: ""
    }

    fun setEndLabel(value: String?) {
        endLabel = value ?: ""
    }

    fun setFloatingLabel(value: String?) {
        floatingLabel = value ?: ""
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        viewRect.apply {
            top = horizon_bottom_margin - textMargin.toInt()
            right = w
            left = 0
            bottom = h
        }
        availableViewWidth = w - circleRadius * 4
        super.onSizeChanged(w, h, oldw, oldh)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawLabels(canvas)
        drawHorizontal(canvas)
        clipCanvas(canvas)
        drawSun(canvas)
    }

    private fun drawLabels(canvas: Canvas) {
        val y = viewRect.bottom - 1f
        drawStartLabel(canvas, y)
        drawEndLabel(canvas, y)
    }

    private fun drawStartLabel(canvas: Canvas, y: Float) {
        val x = viewRect.left + leftMarginLabel
        canvas.drawText(startLabel, x, y, paint)
    }

    private fun drawEndLabel(canvas: Canvas, y: Float) {
        val textWidth = getTextBounds(endLabel).width()
        val x = viewRect.right - textWidth - rightMarginLabel
        canvas.drawText(endLabel, x, y, paint)
    }

    private fun drawHorizontal(canvas: Canvas) {
        val y = viewRect.bottom - horizon_bottom_margin.toFloat()
        val fromX = viewRect.right.toFloat()
        val toX = viewRect.left.toFloat()
        canvas.drawLine(toX, y, fromX, y, paint)
    }

    private fun drawSun(canvas: Canvas) {
        drawCircle(canvas)
        if (sunIsVisible) drawFloatingLabel(canvas)
    }

    private fun clipCanvas(canvas: Canvas) {
        canvas.clipRect(0,0, canvas.width,canvas.height - horizon_bottom_margin)
    }

    fun setPercentProgress(progress: Double) {
        val position = Math.PI + progress * Math.PI
        if (progress > 1 || progress < 0) return clearSun()
        sunCoordinates.x = ((50 + cos(position) * 50) / 100).toFloat()
        sunCoordinates.y = (abs(sin(position) * 100) / 100).toFloat()
        sunIsVisible = true
        invalidate()
    }

    private fun clearSun() {
        sunCoordinates.x = 0f
        sunCoordinates.y = 0f
        sunIsVisible = false
    }

    private fun drawCircle(canvas: Canvas) {
        val sunX = sunCoordinates.x * availableViewWidth
        val sunY = sunCoordinates.y * viewRect.height()
        val cx = sunX + (circleRadius * 2).toFloat()
        val cy = (viewRect.height() - sunY) + circleRadius + textMargin
        val radius = circleRadius.toFloat()
        canvas.drawCircle(cx, cy, radius, paint)
    }

    private fun drawFloatingLabel(canvas: Canvas) {
        val bounds = getTextBounds(floatingLabel)
        val sunX = sunCoordinates.x * availableViewWidth
        val sunY = sunCoordinates.y * viewRect.height()
        val halfTextWidth = bounds.width() / 2
        val doubleTextHeight = bounds.height() * 2
        val sunDiameter = circleRadius * 2f
        val x = (sunX - halfTextWidth).toInt() + sunDiameter
        val y = (viewRect.height() - sunY - doubleTextHeight).toInt() + circleRadius + textMargin
        canvas.drawText(floatingLabel, x, y, paint)
    }

    private fun getTextBounds(text: String): Rect {
        val bounds = Rect()
        paint.getTextBounds(text, 0, text.length, bounds)
        return bounds
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val height = measuredWidth * 0.6
        val newHeight = height.toInt()
        setMeasuredDimension(measuredWidth, newHeight)
    }
}